keps/sig-api-machinery/3962-mutating-admission-policies/README.md

```diff
@@ -0,0 +1,1346 @@
+- Inherit the declarative nature;
+- Existing merging strategy to be leverage without rewritten and the effort of maintaining multiple systems;
+- Support the existing makers/openapi extensions.
+
+Cons:
+- Lack of deletion support (see the unsetting values section for the details and workaround);
+- All operations of SSA require the schemas of the incoming objects, which require the schemas to be available from the discovery endpoint or compiled-in. 
+  - For API extensions types like CRDs, this can cause a delay before the policy can start enforcing requests.
+  
+
+### Phase 1
+
+#### API Shape
+Similar to the `validations` field in ValidatingAdmissionPolicy, a `mutations` field will be defined inside MutatingAdmissionPolicy which allows users to define a list of mutations that apply to the specific resources.
+Each mutation field contains a CEL expression which evaluates to a partially populated object representing a Server Side Apply "apply configuration". The apply configuration is then merged into the request object.
+
+Here is an example of injecting an initContainer.
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: MutatingAdmissionPolicy
+metadata:
+  name: "sidecar-policy.example.com"
+spec:
+  paramKind:
+    group: mutations.example.com
+    kind: Sidecar
+    version: v1
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE"]
+      resources:   ["pods"]
+  matchConditions:
+    - name: does-not-already-have-sidecar
+      expression: "!object.spec.initContainers.exists(ic, ic.name == params.name)"
+  failurePolicy: Fail
+  reinvocationPolicy: IfNeeded
+  mutations:
+    - mutation: >
+        Object{
+          spec: Object.spec{
+            initContainers: [
+              Object.spec.initContainers{
+                name: params.name,
+                image: params.image,
+                args: params.args,
+                restartPolicy: params.restartPolicy
+                // ... other container fields injected here ...
+              }
+            ]
+          }
+        }
+```
+
+For each mutation expression, the expression evaluates to an object that has the same type as the incoming object, and the type alias Object refers to the type (see Type Handling for details).
+
+By using Server Side Apply merge algorithms, schema declarations like
+`x-kubernetes-list-type: map`, that control how a merge is performed, will be
+respected.
+
+However, unlike with server side apply, these mutations will not have a field
+manager. This has important implications in how the merge is performed that will
+be discussed in more detail in the below "Unsetting values" section.
+
+In this example, note that:
+
+- `Object{}`, `Object.spec{}` and similar are CEL object instantiations, and are
+  used to create a subset of the fields of a `Pod`.
+- `oldObject` refers to the state of the object before the mutation is applied.
```

Comment:
Actually, how are you even getting this.  I think admission plugins today see the state of the object after whichever admission plugins were earlier in the chain mutated.  Is this

1. the state of the object after conversion, before any admission started
2. the state of the object when the processing of this mutating admission policy was called
3. the state of the object currently in etcd (creates wouldn't have this)
4. something else?

Reply:
It should be 2 the state of the object when the processing of this mutating admission policy was called. And we could use oldObject to refer to the state of object currently in etcd which aligns with VAP. I'll update shortly